src/state/slices/columnsSlice.tsx

```diff
@@ -0,0 +1,160 @@
+    selectColumn: (state, action: PayloadAction<string>) => {
+      state.selectedColumnIds.push(action.payload);
+    },
+    deselectColumn: (state, action: PayloadAction<string>) => {
+      delete state.sort[action.payload];
+
+      const newSelectedColumnsIds = state.selectedColumnIds.filter(
+        (colId) => colId !== action.payload
+      );
+      state.selectedColumnIds = newSelectedColumnsIds;
+    },
+    reorderColumn: (state, action: PayloadAction<DropResult>) => {
+      const result = action.payload;
+      if (result.destination) {
+        const [reorderedItem] = state.selectedColumnIds.splice(
+          result.source.index,
+          1
+        );
+        state.selectedColumnIds.splice(
+          result.destination?.index,
+          0,
+          reorderedItem
+        );
+      }
+    },
+    changeSort: (
+      state,
+      action: PayloadAction<{ column: string; order: Order | null }>
+    ) => {
+      const { column, order } = action.payload;
+      if (order !== null) {
+        state.sort[column] = order;
+      } else {
+        delete state.sort[column];
+      }
+    },
+    changePage: (state, action: PayloadAction<number>) => {
+      state.page = action.payload;
+    },
+    changeResultsPerPage: (state, action: PayloadAction<number>) => {
+      state.resultsPerPage = action.payload;
+      state.page = 0;
+    },
+    toggleWordWrap: (state, action: PayloadAction<string>) => {
+      if (state.columnStates[action.payload]) {
+        state.columnStates[action.payload].wordWrap =
+          !state.columnStates[action.payload].wordWrap;
+      } else {
+        state.columnStates[action.payload] = { wordWrap: true };
+      }
+    },
+  },
+});
+
+export const {
+  selectColumn,
+  deselectColumn,
+  reorderColumn,
+  changeSort,
+  changePage,
+  changeResultsPerPage,
+  toggleWordWrap,
+} = columnsSlice.actions;
+
+// Other code such as selectors can use the imported `RootState` type
+export const selectColumnStates = (state: RootState) =>
+  state.columns.columnStates;
+export const selectAvailableColumns = (
+  state: RootState,
+  availableColumns: Column[]
+) => availableColumns;
+export const selectSelectedIds = (state: RootState) =>
+  state.columns.selectedColumnIds;
+export const selectSort = (state: RootState) => state.columns.sort;
+export const selectPage = (state: RootState) => state.columns.page;
+export const selectResultsPerPage = (state: RootState) =>
+  state.columns.resultsPerPage;
+
+function arrayEquals(a: string[], b: string[]) {
+  const sortedA = [...a].sort();
+  const sortedB = [...b].sort();
+  return (
+    Array.isArray(sortedA) &&
+    Array.isArray(sortedB) &&
+    sortedA.length === sortedB.length &&
+    sortedA.every((val, index) => val === sortedB[index])
+  );
+}
+
+const selectSelectedIdsIgnoreOrder = createSelector(
+  selectSelectedIds,
+  (selectedIds) => selectedIds,
+  {
+    memoizeOptions: { equalityCheck: arrayEquals },
+  }
+);
```

Comment:
Could you explain a bit more what's going on here? We have this function here which just returns the `selectedIds` (this is just an array of the selected column names, right? Like `['timestamp', 'shotnum', 'activeArea']` that sort of thing?). And we also have `selectSelectedColumns`, which I can see is used in `columnCheckboxes` to filter out those which haven't been ticked.

What's the differences between these two functions and why does the order of the IDs (column names?) matter here? We also have that unit test to check that hidden columns selector ignores the order of `selectedColumnIds`. Why does _that_ matter?

I think I'm just getting confused by all the inclusions of "select". The word "select" itself is conventionally used in Redux to refer to selectors. Then we have our definitions of what columns and stuff have been selected by the user. It's just a headache to get around 😅

Reply:
`selectSelectedIdsIgnoreOrder` is needed as the `selectedIds` state in redux technically stores 2 pieces of info: the selected columns and the column order. You can see that when columns get reordered they affect this array instead of a separate `columnOrder` state - this is so we don't have duplicate state.

However, a side effect in having selected columns and column order derived from the same redux state is that in the case of knowing the selected columns (i.e. what gets passed as `hiddenColumns` to react-table/as `selectedColumns` to `columnCheckboxes`, order doesn't matter and it would cause unnecessary rerendering if it changed when the column order changed. Hence why we have a `selectSelectedIdsIgnoreOrder` selector which uses memoization to sort the array to provide a "stable" array to serve as `selectedColumns`/`hiddenColumns`.

The unit test for `selectHiddenColumns` is to test the memoization and to ensure that this is effective. You can probably test this makes things more efficient by removing the `memoizeOptions` bit in `selectSelectedIdsIgnoreOrder`, and then try and reorder some columns. I had some noticeable lag without the memoization.

Yeah, I'm trying to use the convention which names selectors as `selectPortionOfState`, but separately there's the idea of "selected" columns.

I should probably write portions of this as comments around these specific selectors to explain what's going on. I spent a lot of time working on the selectors and optimising things so I think by the time I had everything working I kinda forgot about commenting it 😓

Reply:
Right, that explains everything amazingly! But yes that was going to be my next ask - please put some of this in comments! Because I know for a fact I'd like to refer back to this information at some point and I presume some others who work on this project down the line will too. 

So put a comment above the unit test in question to say you are testing the memoization part of the function and put some comments somewhere about what the functions do and how this achieves efficiency in rendering please :)

Reply:
@sam-glendenning please read through the comments I added and see if it makes sense & I haven't missed anything :)

Reply:
Comments are good!